src/main.cpp

```diff
@@ -140,23 +145,14 @@ int main(int argc, char *argv[])
 
   if (script.empty())
   {
-    // There should only be 1 non-option argument (the script file)
-    if (optind != argc-1)
-    {
-      usage();
-      return 1;
-    }
+    // Script file
```

Comment:
does this mean that whenever `-e '…'` is omitted: we get a positional parameter for "script path"?

```bash
echo 'BEGIN { printf("I got %d, %s\n", $1, str($2)); }' > mycoolscript.bt
bpftrace mycoolscript.bt 42 "hello"
```

certainly, this way avoids causing any breaking changes, but I think the context-sensitive nature of it makes it a bit surprising.

admittedly, there is some precedent for this; ffmpeg does something similar: the final positional parameter is interpreted as the "output filepath" (if you've not already specified one with `-o …`).

but if we were to copy dtrace, then positional parameters would only ever be "command-line args", and filepath would only ever be an option, -s.

Reply:
ah, I wouldn't want to make filepath available as a positional param $0.

rather, I'd prefer to kill "filepath as a positional param", and instead implement "filepath as an option".

that way, you achieve: "every positional param that you see, is an argument to your script". it removes the "except in the circumstance that you've omitted the -e option, in which case the first positional param has special treatment" behaviour.

Reply:
It would follow the convention of shell scripting ($0 is the name of the script), which is what I'm basing these parameters on in the first place. For now I'm not implementing $0, but we can revisit it later if need be.

Another follow on could be `$#`, for the number of positional parameters, also from shell scripting.

Reply:
After re-reading: I understand now that you've implemented "user-provided positional parameters start from $1 under all circumstances", which is good.

====

And I agree that we don't need to implement $0 right now. I see that bash's convention is to fill it with either the name of the script, or the name used to invoke bash:

> ```
>   Special Parameters
>        0      Expands  to the name of the shell or shell script.  This is set at shell initialization.  If
>               bash is invoked with a file of commands, $0 is set to the name of that  file.   If  bash  is
>               started with the -c option, then $0 is set to the first argument after the string to be exe-
>               cuted, if one is present.  Otherwise, it is set to the file name used  to  invoke  bash,  as
>               given by argument zero.
> ```

```bash
🍔 echo 'echo $0' > test.sh
🍔 bash test.sh
test.sh
🍔 bash -c 'echo $0'
bash
```

====

I was wondering how to document "user may supply filepath as an argument, depending on what options are encountered"; it's important to make it clear to the reader which arguments will be mapped to positional parameters.

GNU bash documents this convention like so:

```
bash [options] [file]
ARGUMENTS
       If  arguments  remain  after  option processing, and neither the -c nor the -s option has been sup-
       plied, the first argument is assumed to be the name of a file containing shell commands.   If  bash
       is  invoked  in  this fashion, $0 is set to the name of the file, and the positional parameters are
       set to the remaining arguments.
```

====

Conclusion: what you've implemented closely follows bash, so LGTM.